reconciler/healthchecks/healthchecks.go

```diff
@@ -231,53 +259,57 @@ func (s *checkStatus) performHealthCheck(serverIP string, healthCheck *types.Rea
 		Timeout:   timeout,
 	}
 
-	serverURL, err := url.Parse(fmt.Sprintf("http://%s:%s%s", serverIP, checkURL.Port(), checkURL.Path))
+	serverURL, err := url.Parse(fmt.Sprintf("http://%s:%s%s", s.serverIP, checkURL.Port(), checkURL.Path))
 	if err != nil {
 		panic(err)
 	}
 
 	resp, err := client.Get(serverURL.String())
 	if err != nil {
 		log.Infof("%s inaccessible: %v", serverURL, err)
-		s.incrementDown(healthCheck)
+		s.incrementDown()
 		return
 	}
 	defer resp.Body.Close()
 	if resp.StatusCode < 200 || 300 <= resp.StatusCode {
 		body, _ := ioutil.ReadAll(resp.Body)
 		log.Infof("%s returned %d: %s", serverURL, resp.StatusCode, string(body))
-		s.incrementDown(healthCheck)
+		s.incrementDown()
 		return
 	}
-	s.incrementUp(healthCheck)
+	s.incrementUp()
 }
 
-func (s *checkStatus) incrementUp(healthCheck *types.RealServer_HealthCheck) {
+func (s *healthCheckState) incrementUp() {
 	s.Lock()
 	defer s.Unlock()
-	switch s.state {
-	case serverDown:
-		s.count++
-		if s.count >= healthCheck.UpThreshold {
-			s.state = serverUp
-			s.count = 0
+	status := s.status
+	switch status.state {
+	case ServerDown:
+		status.count++
+		if status.count >= s.healthCheck.UpThreshold {
+			status.state = ServerUp
+			status.count = 0
+			s.transitionFn(status.state)
 		}
-	case serverUp:
-		s.count = 0
+	case ServerUp:
+		status.count = 0
 	}
 }
 
-func (s *checkStatus) incrementDown(healthCheck *types.RealServer_HealthCheck) {
+func (s *healthCheckState) incrementDown() {
 	s.Lock()
 	defer s.Unlock()
-	switch s.state {
-	case serverDown:
-		s.count = 0
-	case serverUp:
-		s.count++
-		if s.count >= healthCheck.DownThreshold {
-			s.state = serverDown
-			s.count = 0
+	status := s.status
```

Comment:
I originally did that but it's horrible passing in all those arguments, don't you think? It means we're not at the right level of abstraction, it's better to use the struct with the fields in it, grouped nicely, rather than having to add/expand this argument list.

Reply:
I agree about the duplication though.

Reply:
maybe the transition function belongs directly to the state, as this is who actually consumes it when it triggers the change. 
I would say the thresholds may as well, the ones in the healthcheck are only for configuration. 

Then you don't need to pass the parameters.

Reply:
It's still not clear to me the advantage. I'd have to duplicate the threshold values inside the state struct - they already exist at the top level. What's the point?

Reply:
Hmm need to figure out a different approach to your suggestion - since I don't think it's good to rely on the boolean, what if we wanted more than one state?